lib/stats-writer-plugin.js

```diff
@@ -3,6 +3,17 @@
 const INDENT = 2;
 const DEFAULT_TRANSFORM = (data) => JSON.stringify(data, null, INDENT);
 
+// Helper to call a function only once.
+const once = function (fn) {
+  let called = false;
+  return (...args) => {
+    if (!called) {
+      called = true;
+      fn(...args);
+    }
+  };
+};
```

Comment:
It will break watch mode for your plugin

Reply:
Can you help explain this a little more to me? I am using this with the sample repository that @jdelStrother  provided and specifically in testing watch mode which now appears to work correctly.

Without this change, I see multiple invocations of the hooks on every watch change. They add up on each file change and it seems like the wrong thing is happening.

Reply:
It is expected, because developer can remove `dist` directory on each run, **never** use `once` for hooks

Reply:
To explain what I'm seeing a little more, here's what is happening right now in watch mode with current published 1.0.1 version of the plugin in watch mode in https://github.com/jdelStrother/webpack-stats-test repository:

Initial start:

- outer `compiler.hooks.thisCompilation.tap("stats-writer-plugin")` fires
- nested `compilation.hooks.processAssets.tapPromise()` fires

File change #1:

- outer `compiler.hooks.thisCompilation.tap("stats-writer-plugin")` fires
- nested `compilation.hooks.processAssets.tapPromise()` fires
- nested `compilation.hooks.processAssets.tapPromise()` fires

File change #2:

- outer `compiler.hooks.thisCompilation.tap("stats-writer-plugin")` fires
- nested `compilation.hooks.processAssets.tapPromise()` fires
- nested `compilation.hooks.processAssets.tapPromise()` fires
- nested `compilation.hooks.processAssets.tapPromise()` fires

... and so on. 

The reason we get the original `multiple assets emit different content to the same filename` warning in https://github.com/FormidableLabs/webpack-stats-plugin/issues/59 is because I think of this issue and not anything more (which means the `updateAsset()` tweak is more just a nicety than actual core issue solution.)

What my read is that we only need `compiler.hooks.thisCompilation.tap("stats-writer-plugin")` to fire **once** because I need to capture the `compilation` object to manually pass into `compilation.hooks.processAssets` later where it otherwise wouldn't be available.

With the changes in this PR, this it the new behavior:

Initial start:

- outer `compiler.hooks.thisCompilation.tap("stats-writer-plugin")` fires
- nested `compilation.hooks.processAssets.tapPromise()` fires

File change #1:

- nested `compilation.hooks.processAssets.tapPromise()` fires

File change #2:

- nested `compilation.hooks.processAssets.tapPromise()` fires

... and so on. The original `compilation` object captured in initial start is passed to the function in `compilation.hooks.processAssets`.

Does that make sense or is there some reason why we can't capture the `compilation` object initially in `compiler.hooks.thisCompilation.tap("stats-writer-plugin")` and not recapture after?

What I'm thinking if there is some need to update the `compilation` object is to do it in a way that separates `compilation.hooks.processAssets.tapPromise()` from `compiler.hooks.thisCompilation.tap("stats-writer-plugin")` so that `compilation.hooks.processAssets.tapPromise()` continues to only be called once during a straight `webpack` build or during a single file change in `webpack --watch`.

Reply:
Ah, I tried out @jdelStrother's repo with webpack 5 and I am seeing an issue where further hooks aren't updated, which is probably related to `compilation` object. I'm investigating further...

Reply:
just note - https://github.com/FormidableLabs/webpack-stats-plugin/blob/dc4336389d002dc1a17c6b446d84ede13b1586c4/lib/stats-writer-plugin.js#L101 you need to return promise here, because you use `tapPromise`

Reply:
Thanks for the note! `emitStats()` returns a promise.